// @chakra-ui/table/dist/chakra-ui-table.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var system = require('@chakra-ui/system');
var utils = require('@chakra-ui/utils');
var React = require('react');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n["default"] = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespace(React);

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

const TableContainer = /*#__PURE__*/system.forwardRef((props, ref) => {
  const {
    overflow,
    overflowX,
    className,
    ...rest
  } = props;
  return /*#__PURE__*/React__namespace.createElement(system.chakra.div, _extends({
    ref: ref,
    className: utils.cx("chakra-table__container", className)
  }, rest, {
    __css: {
      display: "block",
      whiteSpace: "nowrap",
      WebkitOverflowScrolling: "touch",
      overflowX: overflow ?? overflowX ?? "auto",
      overflowY: "hidden",
      maxWidth: "100%"
    }
  }));
});
const Table = /*#__PURE__*/system.forwardRef((props, ref) => {
  const styles = system.useMultiStyleConfig("Table", props);
  const {
    className,
    ...tableProps
  } = system.omitThemingProps(props);
  return /*#__PURE__*/React__namespace.createElement(system.StylesProvider, {
    value: styles
  }, /*#__PURE__*/React__namespace.createElement(system.chakra.table, _extends({
    role: "table",
    ref: ref,
    __css: styles.table,
    className: utils.cx("chakra-table", className)
  }, tableProps)));
});

if (utils.__DEV__) {
  Table.displayName = "Table";
}

const TableCaption = /*#__PURE__*/system.forwardRef((props, ref) => {
  const {
    placement = "bottom",
    ...rest
  } = props;
  const styles = system.useStyles();
  return /*#__PURE__*/React__namespace.createElement(system.chakra.caption, _extends({}, rest, {
    ref: ref,
    __css: { ...styles.caption,
      captionSide: placement
    }
  }));
});

if (utils.__DEV__) {
  TableCaption.displayName = "TableCaption";
}

const Thead = /*#__PURE__*/system.forwardRef((props, ref) => {
  const styles = system.useStyles();
  return /*#__PURE__*/React__namespace.createElement(system.chakra.thead, _extends({}, props, {
    ref: ref,
    __css: styles.thead
  }));
});
const Tbody = /*#__PURE__*/system.forwardRef((props, ref) => {
  const styles = system.useStyles();
  return /*#__PURE__*/React__namespace.createElement(system.chakra.tbody, _extends({}, props, {
    ref: ref,
    __css: styles.tbody
  }));
});
const Tfoot = /*#__PURE__*/system.forwardRef((props, ref) => {
  const styles = system.useStyles();
  return /*#__PURE__*/React__namespace.createElement(system.chakra.tfoot, _extends({}, props, {
    ref: ref,
    __css: styles.tfoot
  }));
});
const Th = /*#__PURE__*/system.forwardRef(({
  isNumeric,
  ...rest
}, ref) => {
  const styles = system.useStyles();
  return /*#__PURE__*/React__namespace.createElement(system.chakra.th, _extends({}, rest, {
    ref: ref,
    __css: styles.th,
    "data-is-numeric": isNumeric
  }));
});
const Tr = /*#__PURE__*/system.forwardRef((props, ref) => {
  const styles = system.useStyles();
  return /*#__PURE__*/React__namespace.createElement(system.chakra.tr, _extends({
    role: "row"
  }, props, {
    ref: ref,
    __css: styles.tr
  }));
});
const Td = /*#__PURE__*/system.forwardRef(({
  isNumeric,
  ...rest
}, ref) => {
  const styles = system.useStyles();
  return /*#__PURE__*/React__namespace.createElement(system.chakra.td, _extends({
    role: "gridcell"
  }, rest, {
    ref: ref,
    __css: styles.td,
    "data-is-numeric": isNumeric
  }));
});

exports.Table = Table;
exports.TableCaption = TableCaption;
exports.TableContainer = TableContainer;
exports.Tbody = Tbody;
exports.Td = Td;
exports.Tfoot = Tfoot;
exports.Th = Th;
exports.Thead = Thead;
exports.Tr = Tr;
