// @clerk/nextjs/dist/esm/pages/ClerkProvider.js
import { ClerkProvider as ReactClerkProvider } from "@clerk/clerk-react";
import { setClerkJsLoadingErrorPackageName, setErrorThrowerOptions } from "@clerk/clerk-react/internal";
import { useRouter } from "next/router";
import React from "react";
import { useSafeLayoutEffect } from "../client-boundary/hooks/useSafeLayoutEffect";
import { ClerkNextOptionsProvider } from "../client-boundary/NextOptionsContext";
import { ClerkJSScript } from "../utils/clerk-js-script";
import { invalidateNextRouterCache } from "../utils/invalidateNextRouterCache";
import { mergeNextClerkPropsWithEnv } from "../utils/mergeNextClerkPropsWithEnv";
import { removeBasePath } from "../utils/removeBasePath";
setErrorThrowerOptions({ packageName: "@clerk/nextjs" });
setClerkJsLoadingErrorPackageName("@clerk/nextjs");
const NEXT_WINDOW_HISTORY_SUPPORT_VERSION = "14.1.0";
const useNextRouter = () => {
  const router = useRouter();
  const canUseWindowHistoryAPIs = typeof window !== "undefined" && window.next && window.next.version >= NEXT_WINDOW_HISTORY_SUPPORT_VERSION;
  return {
    mode: "path",
    name: "NextRouter",
    push: (path) => router.push(path),
    replace: (path) => canUseWindowHistoryAPIs ? window.history.replaceState(null, "", path) : router.replace(path),
    shallowPush(path) {
      canUseWindowHistoryAPIs ? window.history.pushState(null, "", path) : router.push(path, {});
    },
    pathname: () => window.location.pathname,
    searchParams: () => new URLSearchParams(window.location.search)
  };
};
function ClerkProvider({ children, ...props }) {
  var _a;
  const { __unstable_invokeMiddlewareOnAuthStateChange = true } = props;
  const { push, replace } = useRouter();
  const clerkRouter = useNextRouter();
  ReactClerkProvider.displayName = "ReactClerkProvider";
  useSafeLayoutEffect(() => {
    window.__unstable__onBeforeSetActive = invalidateNextRouterCache;
  }, []);
  useSafeLayoutEffect(() => {
    window.__unstable__onAfterSetActive = () => {
      if (__unstable_invokeMiddlewareOnAuthStateChange) {
        void push(window.location.href);
      }
    };
  }, []);
  const navigate = (to) => push(removeBasePath(to));
  const replaceNavigate = (to) => replace(removeBasePath(to));
  const mergedProps = mergeNextClerkPropsWithEnv({
    ...props,
    __experimental_router: clerkRouter,
    routerPush: navigate,
    routerReplace: replaceNavigate
  });
  const initialState = ((_a = props.authServerSideProps) == null ? void 0 : _a.__clerk_ssr_state) || props.__clerk_ssr_state;
  return /* @__PURE__ */ React.createElement(ClerkNextOptionsProvider, { options: mergedProps }, /* @__PURE__ */ React.createElement(
    ReactClerkProvider,
    {
      ...mergedProps,
      initialState
    },
    /* @__PURE__ */ React.createElement(ClerkJSScript, { router: "pages" }),
    children
  ));
}
export {
  ClerkProvider
};
