// @chakra-ui/provider/dist/chakra-ui-provider.esm.js
import CSSReset from '@chakra-ui/css-reset';
import { PortalManager } from '@chakra-ui/portal';
import { ThemeProvider, ColorModeProvider, GlobalStyle } from '@chakra-ui/system';
import { EnvironmentProvider } from '@chakra-ui/react-env';
import * as React from 'react';
import { IdProvider } from '@chakra-ui/hooks';

/**
 * The global provider that must be added to make all Chakra components
 * work correctly
 */
const ChakraProvider = props => {
  const {
    children,
    colorModeManager,
    portalZIndex,
    resetCSS = true,
    theme = {},
    environment,
    cssVarsRoot
  } = props;

  const _children = /*#__PURE__*/React.createElement(EnvironmentProvider, {
    environment: environment
  }, children);

  return /*#__PURE__*/React.createElement(IdProvider, null, /*#__PURE__*/React.createElement(ThemeProvider, {
    theme: theme,
    cssVarsRoot: cssVarsRoot
  }, /*#__PURE__*/React.createElement(ColorModeProvider, {
    colorModeManager: colorModeManager,
    options: theme.config
  }, resetCSS && /*#__PURE__*/React.createElement(CSSReset, null), /*#__PURE__*/React.createElement(GlobalStyle, null), portalZIndex ? /*#__PURE__*/React.createElement(PortalManager, {
    zIndex: portalZIndex
  }, _children) : _children)));
};

export { ChakraProvider };
