// @chakra-ui/editable/dist/chakra-ui-editable.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var system = require('@chakra-ui/system');
var utils = require('@chakra-ui/utils');
var reactUtils = require('@chakra-ui/react-utils');
var React = require('react');
var hooks = require('@chakra-ui/hooks');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n["default"] = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespace(React);

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

/**
 * React hook for managing the inline renaming of some text.
 *
 * @see Docs https://chakra-ui.com/editable
 */
function useEditable(props = {}) {
  const {
    onChange: onChangeProp,
    onCancel: onCancelProp,
    onSubmit: onSubmitProp,
    value: valueProp,
    isDisabled,
    defaultValue,
    startWithEditView,
    isPreviewFocusable = true,
    submitOnBlur = true,
    selectAllOnFocus = true,
    placeholder,
    onEdit: onEditProp,
    ...htmlProps
  } = props;
  const defaultIsEditing = Boolean(startWithEditView && !isDisabled);
  const [isEditing, setIsEditing] = React.useState(defaultIsEditing);
  const [value, setValue] = hooks.useControllableState({
    defaultValue: defaultValue || "",
    value: valueProp,
    onChange: onChangeProp
  });
  /**
   * Keep track of the previous value, so if users
   * presses `cancel`, we can revert to it.
   */

  const [prevValue, setPrevValue] = React.useState(value);
  /**
   * Ref to help focus the input in edit mode
   */

  const inputRef = React.useRef(null);
  const previewRef = React.useRef(null);
  const editButtonRef = React.useRef(null);
  const cancelButtonRef = React.useRef(null);
  const submitButtonRef = React.useRef(null);
  hooks.useFocusOnPointerDown({
    ref: inputRef,
    enabled: isEditing,
    elements: [cancelButtonRef, submitButtonRef]
  });
  const isInteractive = !isEditing || !isDisabled;
  hooks.useUpdateEffect(() => {
    if (!isEditing) {
      utils.focus(editButtonRef.current);
      return;
    }

    utils.focus(inputRef.current, {
      selectTextIfInput: selectAllOnFocus
    });
    onEditProp?.();
  }, [isEditing, onEditProp, selectAllOnFocus]);
  const onEdit = React.useCallback(() => {
    if (isInteractive) {
      setIsEditing(true);
    }
  }, [isInteractive]);
  const onCancel = React.useCallback(() => {
    setIsEditing(false);
    setValue(prevValue);
    onCancelProp?.(prevValue);
  }, [onCancelProp, setValue, prevValue]);
  const onSubmit = React.useCallback(() => {
    setIsEditing(false);
    setPrevValue(value);
    onSubmitProp?.(value);
  }, [value, onSubmitProp]);
  const onChange = React.useCallback(event => {
    setValue(event.target.value);
  }, [setValue]);
  const onKeyDown = React.useCallback(event => {
    const eventKey = utils.normalizeEventKey(event);
    const keyMap = {
      Escape: onCancel,
      Enter: event => {
        if (!event.shiftKey && !event.metaKey) {
          onSubmit();
        }
      }
    };
    const action = keyMap[eventKey];

    if (action) {
      event.preventDefault();
      action(event);
    }
  }, [onCancel, onSubmit]);
  const isValueEmpty = utils.isEmpty(value);
  const onBlur = React.useCallback(event => {
    const relatedTarget = utils.getRelatedTarget(event);
    const targetIsCancel = utils.contains(cancelButtonRef.current, relatedTarget);
    const targetIsSubmit = utils.contains(submitButtonRef.current, relatedTarget);
    const isValidBlur = !targetIsCancel && !targetIsSubmit;

    if (isValidBlur && submitOnBlur) {
      onSubmit();
    }
  }, [submitOnBlur, onSubmit]);
  const getPreviewProps = React.useCallback((props = {}, ref = null) => {
    const tabIndex = isInteractive && isPreviewFocusable ? 0 : undefined;
    return { ...props,
      ref: reactUtils.mergeRefs(ref, previewRef),
      children: isValueEmpty ? placeholder : value,
      hidden: isEditing,
      "aria-disabled": utils.ariaAttr(isDisabled),
      tabIndex,
      onFocus: utils.callAllHandlers(props.onFocus, onEdit)
    };
  }, [isDisabled, isEditing, isInteractive, isPreviewFocusable, isValueEmpty, onEdit, placeholder, value]);
  const getInputProps = React.useCallback((props = {}, ref = null) => ({ ...props,
    hidden: !isEditing,
    placeholder,
    ref: reactUtils.mergeRefs(ref, inputRef),
    disabled: isDisabled,
    "aria-disabled": utils.ariaAttr(isDisabled),
    value,
    onBlur: utils.callAllHandlers(props.onBlur, onBlur),
    onChange: utils.callAllHandlers(props.onChange, onChange),
    onKeyDown: utils.callAllHandlers(props.onKeyDown, onKeyDown)
  }), [isDisabled, isEditing, onBlur, onChange, onKeyDown, placeholder, value]);
  const getEditButtonProps = React.useCallback((props = {}, ref = null) => ({
    "aria-label": "Edit",
    ...props,
    type: "button",
    onClick: utils.callAllHandlers(props.onClick, onEdit),
    ref: reactUtils.mergeRefs(ref, editButtonRef)
  }), [onEdit]);
  const getSubmitButtonProps = React.useCallback((props = {}, ref = null) => ({ ...props,
    "aria-label": "Submit",
    ref: reactUtils.mergeRefs(submitButtonRef, ref),
    type: "button",
    onClick: utils.callAllHandlers(props.onClick, onSubmit)
  }), [onSubmit]);
  const getCancelButtonProps = React.useCallback((props = {}, ref = null) => ({
    "aria-label": "Cancel",
    id: "cancel",
    ...props,
    ref: reactUtils.mergeRefs(cancelButtonRef, ref),
    type: "button",
    onClick: utils.callAllHandlers(props.onClick, onCancel)
  }), [onCancel]);
  return {
    isEditing,
    isDisabled,
    isValueEmpty,
    value,
    onEdit,
    onCancel,
    onSubmit,
    getPreviewProps,
    getInputProps,
    getEditButtonProps,
    getSubmitButtonProps,
    getCancelButtonProps,
    htmlProps
  };
}

const [EditableProvider, useEditableContext] = reactUtils.createContext({
  name: "EditableContext",
  errorMessage: "useEditableContext: context is undefined. Seems you forgot to wrap the editable components in `<Editable />`"
});

/**
 * Editable
 *
 * The wrapper that provides context and logic for all editable
 * components. It renders a `div`
 */
const Editable = /*#__PURE__*/system.forwardRef((props, ref) => {
  const styles = system.useMultiStyleConfig("Editable", props);
  const ownProps = system.omitThemingProps(props);
  const {
    htmlProps,
    ...context
  } = useEditable(ownProps);
  const {
    isEditing,
    onSubmit,
    onCancel,
    onEdit
  } = context;

  const _className = utils.cx("chakra-editable", props.className);

  const children = utils.runIfFn(props.children, {
    isEditing,
    onSubmit,
    onCancel,
    onEdit
  });
  return /*#__PURE__*/React__namespace.createElement(EditableProvider, {
    value: context
  }, /*#__PURE__*/React__namespace.createElement(system.StylesProvider, {
    value: styles
  }, /*#__PURE__*/React__namespace.createElement(system.chakra.div, _extends({
    ref: ref
  }, htmlProps, {
    className: _className
  }), children)));
});

if (utils.__DEV__) {
  Editable.displayName = "Editable";
}

const commonStyles = {
  fontSize: "inherit",
  fontWeight: "inherit",
  textAlign: "inherit",
  bg: "transparent"
};

/**
 * EditablePreview
 *
 * The `span` used to display the final value, in the `preview` mode
 */
const EditablePreview = /*#__PURE__*/system.forwardRef((props, ref) => {
  const {
    getPreviewProps
  } = useEditableContext();
  const styles = system.useStyles();
  const previewProps = getPreviewProps(props, ref);

  const _className = utils.cx("chakra-editable__preview", props.className);

  return /*#__PURE__*/React__namespace.createElement(system.chakra.span, _extends({}, previewProps, {
    __css: {
      cursor: "text",
      display: "inline-block",
      ...commonStyles,
      ...styles.preview
    },
    className: _className
  }));
});

if (utils.__DEV__) {
  EditablePreview.displayName = "EditablePreview";
}

/**
 * EditableInput
 *
 * The input used in the `edit` mode
 */
const EditableInput = /*#__PURE__*/system.forwardRef((props, ref) => {
  const {
    getInputProps
  } = useEditableContext();
  const styles = system.useStyles();
  const inputProps = getInputProps(props, ref);

  const _className = utils.cx("chakra-editable__input", props.className);

  return /*#__PURE__*/React__namespace.createElement(system.chakra.input, _extends({}, inputProps, {
    __css: {
      outline: 0,
      ...commonStyles,
      ...styles.input
    },
    className: _className
  }));
});

if (utils.__DEV__) {
  EditableInput.displayName = "EditableInput";
}
/**
 * React hook use to gain access to the editable state and actions.
 */


function useEditableState() {
  const {
    isEditing,
    onSubmit,
    onCancel,
    onEdit,
    isDisabled
  } = useEditableContext();
  return {
    isEditing,
    onSubmit,
    onCancel,
    onEdit,
    isDisabled
  };
}
/**
 * React hook use to create controls for the editable component
 */

function useEditableControls() {
  const {
    isEditing,
    getEditButtonProps,
    getCancelButtonProps,
    getSubmitButtonProps
  } = useEditableContext();
  return {
    isEditing,
    getEditButtonProps,
    getCancelButtonProps,
    getSubmitButtonProps
  };
}

exports.Editable = Editable;
exports.EditableInput = EditableInput;
exports.EditablePreview = EditablePreview;
exports.useEditable = useEditable;
exports.useEditableControls = useEditableControls;
exports.useEditableState = useEditableState;
