// @chakra-ui/portal/dist/chakra-ui-portal.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var utils = require('@chakra-ui/utils');
var reactUtils = require('@chakra-ui/react-utils');
var React = require('react');
var hooks = require('@chakra-ui/hooks');
var reactDom = require('react-dom');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n["default"] = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespace(React);

const [PortalManagerContextProvider, usePortalManager] = reactUtils.createContext({
  strict: false,
  name: "PortalManagerContext"
});
function PortalManager(props) {
  const {
    children,
    zIndex
  } = props;
  return /*#__PURE__*/React__namespace.createElement(PortalManagerContextProvider, {
    value: {
      zIndex
    }
  }, children);
}

if (utils.__DEV__) {
  PortalManager.displayName = "PortalManager";
}

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

const [PortalContextProvider, usePortalContext] = reactUtils.createContext({
  strict: false,
  name: "PortalContext"
});
const PORTAL_CLASSNAME = "chakra-portal";
const PORTAL_SELECTOR = `.chakra-portal`;

const Container = props => /*#__PURE__*/React__namespace.createElement("div", {
  className: "chakra-portal-zIndex",
  style: {
    position: "absolute",
    zIndex: props.zIndex,
    top: 0,
    left: 0,
    right: 0 // NB: Don't add `bottom: 0`, it makes the entire app unusable
    // @see https://github.com/chakra-ui/chakra-ui/issues/3201

  }
}, props.children);
/**
 * Portal that uses `document.body` as container
 */


const DefaultPortal = props => {
  const {
    appendToParentPortal,
    children
  } = props;
  const tempNode = React__namespace.useRef(null);
  const portal = React__namespace.useRef(null);
  const forceUpdate = hooks.useForceUpdate();
  const parentPortal = usePortalContext();
  const manager = usePortalManager();
  hooks.useSafeLayoutEffect(() => {
    if (!tempNode.current) return;
    const doc = tempNode.current.ownerDocument;
    const host = appendToParentPortal ? parentPortal ?? doc.body : doc.body;
    if (!host) return;
    portal.current = doc.createElement("div");
    portal.current.className = PORTAL_CLASSNAME;
    host.appendChild(portal.current);
    forceUpdate();
    const portalNode = portal.current;
    return () => {
      if (host.contains(portalNode)) {
        host.removeChild(portalNode);
      }
    };
  }, []);

  const _children = manager?.zIndex ? /*#__PURE__*/React__namespace.createElement(Container, {
    zIndex: manager?.zIndex
  }, children) : children;

  return portal.current ? /*#__PURE__*/reactDom.createPortal( /*#__PURE__*/React__namespace.createElement(PortalContextProvider, {
    value: portal.current
  }, _children), portal.current) : /*#__PURE__*/React__namespace.createElement("span", {
    ref: tempNode
  });
};

/**
 * Portal that uses a custom container
 */
const ContainerPortal = props => {
  const {
    children,
    containerRef,
    appendToParentPortal
  } = props;
  const containerEl = containerRef.current;
  const host = containerEl ?? (utils.isBrowser ? document.body : undefined);
  const portal = React__namespace.useMemo(() => {
    const node = containerEl?.ownerDocument.createElement("div");
    if (node) node.className = PORTAL_CLASSNAME;
    return node;
  }, [containerEl]);
  const forceUpdate = hooks.useForceUpdate();
  hooks.useSafeLayoutEffect(() => {
    forceUpdate();
  }, []);
  hooks.useSafeLayoutEffect(() => {
    if (!portal || !host) return;
    host.appendChild(portal);
    return () => {
      host.removeChild(portal);
    };
  }, [portal, host]);

  if (host && portal) {
    return /*#__PURE__*/reactDom.createPortal( /*#__PURE__*/React__namespace.createElement(PortalContextProvider, {
      value: appendToParentPortal ? portal : null
    }, children), portal);
  }

  return null;
};

/**
 * Portal
 *
 * Declarative component used to render children into a DOM node
 * that exists outside the DOM hierarchy of the parent component.
 *
 * @see Docs https://chakra-ui.com/portal
 */
function Portal(props) {
  const {
    containerRef,
    ...rest
  } = props;
  return containerRef ? /*#__PURE__*/React__namespace.createElement(ContainerPortal, _extends({
    containerRef: containerRef
  }, rest)) : /*#__PURE__*/React__namespace.createElement(DefaultPortal, rest);
}
Portal.defaultProps = {
  appendToParentPortal: true
};
Portal.className = PORTAL_CLASSNAME;
Portal.selector = PORTAL_SELECTOR;

if (utils.__DEV__) {
  Portal.displayName = "Portal";
}

exports.Portal = Portal;
exports.PortalManager = PortalManager;
exports.usePortalManager = usePortalManager;
