// @chakra-ui/anatomy/dist/chakra-ui-anatomy.esm.js
import { anatomy } from '@chakra-ui/theme-tools';

/**
 * **Accordion anatomy**
 * - Item: the accordion item contains the button and panel
 * - Button: the button is the trigger for the panel
 * - Panel: the panel is the content of the accordion item
 * - Icon: the expanded/collapsed icon
 */

const accordionAnatomy = anatomy("accordion").parts("container", "item", "button", "panel").extend("icon");
/**
 * **Alert anatomy**
 * - Title: the alert's title
 * - Description: the alert's description
 * - Icon: the alert's icon
 */

const alertAnatomy = anatomy("alert").parts("title", "description", "container").extend("icon");
/**
 * **Avatar anatomy**
 * - Container: the container for the avatar
 * - Label: the avatar initials text
 * - Excess Label: the label or text that represents excess avatar count.
 * Typically used in avatar groups.
 * - Group: the container for the avatar group
 */

const avatarAnatomy = anatomy("avatar").parts("label", "badge", "container").extend("excessLabel", "group");
/**
 * **Breadcrumb anatomy**
 * - Item: the container for a breadcrumb item
 * - Link: the element that represents the breadcrumb link
 * - Container: the container for the breadcrumb items
 * - Separator: the separator between breadcrumb items
 */

const breadcrumbAnatomy = anatomy("breadcrumb").parts("link", "item", "container").extend("separator");
const buttonAnatomy = anatomy("button").parts();
const checkboxAnatomy = anatomy("checkbox").parts("control", "icon", "container").extend("label");
const circularProgressAnatomy = anatomy("progress").parts("track", "filledTrack").extend("label");
const drawerAnatomy = anatomy("drawer").parts("overlay", "dialogContainer", "dialog").extend("header", "closeButton", "body", "footer");
const editableAnatomy = anatomy("editable").parts("preview", "input");
const formAnatomy = anatomy("form").parts("container", "requiredIndicator", "helperText");
const formErrorAnatomy = anatomy("formError").parts("text", "icon");
const inputAnatomy = anatomy("input").parts("addon", "field", "element");
const listAnatomy = anatomy("list").parts("container", "item", "icon");
const menuAnatomy = anatomy("menu").parts("button", "list", "item").extend("groupTitle", "command", "divider");
const modalAnatomy = anatomy("modal").parts("overlay", "dialogContainer", "dialog").extend("header", "closeButton", "body", "footer");
const numberInputAnatomy = anatomy("numberinput").parts("root", "field", "stepperGroup", "stepper");
const pinInputAnatomy = anatomy("pininput").parts("field");
const popoverAnatomy = anatomy("popover").parts("content", "header", "body", "footer").extend("popper", "arrow", "closeButton");
const progressAnatomy = anatomy("progress").parts("label", "filledTrack", "track");
const radioAnatomy = anatomy("radio").parts("container", "control", "label");
const selectAnatomy = anatomy("select").parts("field", "icon");
const sliderAnatomy = anatomy("slider").parts("container", "track", "thumb", "filledTrack");
const statAnatomy = anatomy("stat").parts("container", "label", "helpText", "number", "icon");
const switchAnatomy = anatomy("switch").parts("container", "track", "thumb");
const tableAnatomy = anatomy("table").parts("table", "thead", "tbody", "tr", "th", "td", "tfoot", "caption");
const tabsAnatomy = anatomy("tabs").parts("root", "tab", "tablist", "tabpanel", "tabpanels", "indicator");
/**
 * **Tag anatomy**
 * - Container: the container for the tag
 * - Label: the text content of the tag
 * - closeButton: the close button for the tag
 */

const tagAnatomy = anatomy("tag").parts("container", "label", "closeButton");

export { accordionAnatomy, alertAnatomy, avatarAnatomy, breadcrumbAnatomy, buttonAnatomy, checkboxAnatomy, circularProgressAnatomy, drawerAnatomy, editableAnatomy, formAnatomy, formErrorAnatomy, inputAnatomy, listAnatomy, menuAnatomy, modalAnatomy, numberInputAnatomy, pinInputAnatomy, popoverAnatomy, progressAnatomy, radioAnatomy, selectAnatomy, sliderAnatomy, statAnatomy, switchAnatomy, tableAnatomy, tabsAnatomy, tagAnatomy };
